ws-client/src/client.rs

```diff
@@ -90,6 +96,62 @@ pub struct WsClient {
 	error: Mutex<ErrorFromBack>,
 	/// Request timeout
 	request_timeout: Option<Duration>,
+	/// Request ID manager.
+	id_guard: RequestIdGuard,
+}
+
+#[derive(Debug)]
+struct RequestIdGuard {
+	// Current pending requests.
+	current_pending: AtomicUsize,
+	/// Max concurrent pending requests allowed.
+	max_concurrent_requests: usize,
+	/// Get the next request ID.
+	current_id: AtomicU64,
+}
+
+impl RequestIdGuard {
+	fn new(limit: usize) -> Self {
+		Self { current_pending: AtomicUsize::new(0), max_concurrent_requests: limit, current_id: AtomicU64::new(0) }
+	}
+
+	fn get_slot(&self) -> Result<(), Error> {
+		if self.current_pending.load(Ordering::Relaxed) >= self.max_concurrent_requests {
+			Err(Error::MaxSlotsExceeded)
+		} else {
+			// NOTE: `fetch_add` wraps on overflow but that can't occur because `current_pending` is checked above.
+			self.current_pending.fetch_add(1, Ordering::Relaxed);
+			Ok(())
+		}
+	}
+
+	/// Attempts to get the next request ID.
+	///
+	/// Fails if request limit has been exceeded.
+	fn next_request_id(&self) -> Result<u64, Error> {
+		self.get_slot()?;
+		let id = self.current_id.fetch_add(1, Ordering::Relaxed);
+		Ok(id)
+	}
+
+	/// Attempts to get the `n` number next IDs that only counts as one request.
+	///
+	/// Fails if request limit has been exceeded.
+	fn next_request_ids(&self, len: usize) -> Result<Vec<u64>, Error> {
```

Comment:
todo: SmallVec?!

Reply:
imo, not before we have benchmarks proving it's worth it.

Reply:
yepp, we can add an issue for that this PR is already too big :(